### PR TITLE
Improve DAG rendering

### LINKED ISSUE #1056: Improve DAG rendering
## Scenario 
Nextflow supports the rendering of the workflow DAG either in html and Graphviz DOT, which in turns allows the generation in several other formats including pdf, svg, png, etc. 

However the current implementation provides a very basic pictorial representational and does not clearly represents some important elements such as inputs and outputs. See an example below:

dag

Moreover it does not allows the DAG rending without fully execute the workflow application.  

## Deliverable  

The goal of this enhancement is to improve the Nextflow DAG rendering so that: 
1. inputs and outputs are grouped together and more easily identifiable
2. use a color code to better highlight different workflow components 
3. render sub-workflows and custom function/operators implemented in the new dsl 
4. allow interactive navigation of the graph when the DAG is rendered in HTML format.
5. report the number of executed tasks for each node (see #1025)
6. add the ability to render the DAG without the need to run the workflow application (dry run mode)

The resulting graph should be compatible at pictorial level with the DAG generated by the [CWL viewer](https://view.commonwl.org/), for example: 

screenshot 2019-03-03 at 16 44 47

### PR DESCRIPTION
**Pull request by bentsherman**
Close #1056 #1543 #3203 

- Render subworkflows as subgraphs
- Add subgraphs for workflow inputs and outputs
- Add `dag.verbose` option to toggle operators (disabled by default)
- Add `dag.depth` option to control the level of detail

Notes:

- I used wrapper classes around `DAG.Vertex` so that I could store the inputs and outputs for each node. The `DAG` class has all edges in a list instead of a hash map, so not ideal for graph traversal. Maybe the `DAG` class could do this from the beginning, but I haven't looked into it yet.

- I used the fully qualified process names to construct a "node tree" that encodes information about subworkflows. Since operators aren't associated to any subworkflow in the `DAG`, I try to infer it based on its neighboring processes. Maybe the `DAG` could save this information when the operator is added, but the heuristic works pretty well.

- Workflow inputs and outputs are not annotated very well. I can infer some inputs from the outgoing channel name, but the output nodes are always empty. Might be able to grab the output name from the process/workflow `emit` for now.

- Large pipelines like `rnaseq` and `sarek` tend to have lots of little outputs which makes the diagram super long. It would be nice to condense this somehow while still showing something for the workflow outputs.

- Currently specific to Mermaid, but I should be able to make it work for all output formats by moving the pre-rendering logic to `DagRenderer`. Then each renderer will need to operate on the "node tree" instead of the DAG.

Testing:

Run any pipeline with `-preview`, for example:
```bash
../launch.sh run nf-core/fetchngs -profile test --outdir results -preview -with-dag
```

Use `dag.depth` and `dag.verbose` to control the level of detail. For large pipelines like `rnaseq` or `sarek`, I recommend that you keep `verbose=false` and limit the depth to 2 or 3.

### CODE REVIEW

**Comment by bentsherman** [comment]
Here are some examples with `nf-core/fetchngs`.

depth=0, verbose=false
```mermaid
flowchart TD
    subgraph inputs
    v0[Channel.from]
    v5[fields]
    v27[certificate]
    v31[certificate]
    v39[pipeline]
    v40[strandedness]
    v41[mapping_fields]
    end
    v6([NFCORE_FETCHNGS])
    subgraph outputs
    v20[ ]
    v46[ ]
    v49[ ]
    v54[ ]
    v55[ ]
    v56[ ]
    end
    v0 --> v6
    v5 --> v6
    v6 --> v20
    v27 --> v6
    v31 --> v6
    v39 --> v6
    v40 --> v6
    v41 --> v6
    v6 --> v46
    v6 --> v49
    v6 --> v54
    v6 --> v55
    v6 --> v56
```

depth=2, verbose=false
```mermaid
flowchart TD
    subgraph inputs
    v0[Channel.from]
    v5[fields]
    v27[certificate]
    v31[certificate]
    v39[pipeline]
    v40[strandedness]
    v41[mapping_fields]
    end
    subgraph NFCORE_FETCHNGS
    subgraph SRA
    v6([SRA_IDS_TO_RUNINFO])
    v9([SRA_RUNINFO_TO_FTP])
    v19([SRA_FASTQ_FTP])
    v25([FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS])
    v42([SRA_TO_SAMPLESHEET])
    v45([SRA_MERGE_SAMPLESHEET])
    v48([MULTIQC_MAPPINGS_CONFIG])
    v53([CUSTOM_DUMPSOFTWAREVERSIONS])
    v1(( ))
    v7(( ))
    v12(( ))
    v37(( ))
    v43(( ))
    v44(( ))
    end
    end
    subgraph outputs
    v20[ ]
    v46[ ]
    v49[ ]
    v54[ ]
    v55[ ]
    v56[ ]
    end
    v0 --> v1
    v5 --> v6
    v1 --> v6
    v6 --> v9
    v6 --> v7
    v9 --> v7
    v9 --> v12
    v12 --> v19
    v19 --> v20
    v19 --> v7
    v19 --> v37
    v27 --> v25
    v31 --> v25
    v39 --> v42
    v40 --> v42
    v41 --> v42
    v37 --> v42
    v42 --> v43
    v42 --> v44
    v43 --> v45
    v44 --> v45
    v45 --> v46
    v45 --> v48
    v45 --> v7
    v48 --> v49
    v48 --> v7
    v7 --> v53
    v53 --> v56
    v53 --> v55
    v53 --> v54
    v25 --> v7
    v12 --> v25
    v25 --> v37
```

depth=3, verbose=true
```mermaid
flowchart TD
    subgraph inputs
    v0[Channel.from]
    v4[Channel.empty]
    v5[fields]
    v24[Channel.empty]
    v27[certificate]
    v31[certificate]
    v39[pipeline]
    v40[strandedness]
    v41[mapping_fields]
    end
    subgraph NFCORE_FETCHNGS
    subgraph SRA
    v1([splitCsv])
    v2([map])
    v3([unique])
    v6([SRA_IDS_TO_RUNINFO])
    v7([first])
    v8([mix])
    v9([SRA_RUNINFO_TO_FTP])
    v10([first])
    v11([mix])
    v12([splitCsv])
    v13([map])
    v14([unique])
    v15([first])
    v16([mix])
    v17([map])
    v18([branch])
    v19([SRA_FASTQ_FTP])
    v21([first])
    v22([mix])
    v23([map])
    subgraph FASTQ_DOWNLOAD_PREFETCH_FASTERQDUMP_SRATOOLS
    v25([CUSTOM_SRATOOLSNCBISETTINGS])
    v28([SRATOOLS_PREFETCH])
    v32([SRATOOLS_FASTERQDUMP])
    end
    v26([mix])
    v29([first])
    v30([mix])
    v33([first])
    v34([mix])
    v35([first])
    v36([mix])
    v37([mix])
    v38([map])
    v42([SRA_TO_SAMPLESHEET])
    v43([collect])
    v44([collect])
    v45([SRA_MERGE_SAMPLESHEET])
    v47([mix])
    v48([MULTIQC_MAPPINGS_CONFIG])
    v50([mix])
    v51([unique])
    v52([collectFile])
    v53([CUSTOM_DUMPSOFTWAREVERSIONS])
    end
    end
    subgraph outputs
    v20[ ]
    v46[ ]
    v49[ ]
    v54[ ]
    v55[ ]
    v56[ ]
    end
    v0 --> v1
    v1 --> v2
    v2 --> v3
    v3 --> v6
    v4 --> v8
    v5 --> v6
    v6 --> v9
    v6 --> v7
    v7 --> v8
    v8 --> v11
    v9 --> v12
    v9 --> v10
    v9 --> v15
    v10 --> v11
    v11 --> v16
    v12 --> v13
    v13 --> v14
    v14 --> v17
    v15 --> v16
    v16 --> v22
    v17 --> v18
    v18 --> v23
    v18 --> v19
    v19 --> v37
    v19 --> v20
    v19 --> v21
    v21 --> v22
    v22 --> v36
    v23 --> v28
    v24 --> v26
    v25 --> v28
    v25 --> v26
    v25 --> v32
    v26 --> v30
    v27 --> v28
    v28 --> v32
    v28 --> v29
    v29 --> v30
    v30 --> v34
    v31 --> v32
    v32 --> v37
    v32 --> v33
    v33 --> v34
    v34 --> v35
    v35 --> v36
    v36 --> v47
    v37 --> v38
    v38 --> v42
    v39 --> v42
    v40 --> v42
    v41 --> v42
    v42 --> v43
    v42 --> v44
    v43 --> v45
    v44 --> v45
    v45 --> v46
    v45 --> v48
    v45 --> v47
    v47 --> v50
    v48 --> v49
    v48 --> v50
    v50 --> v51
    v51 --> v52
    v52 --> v53
    v53 --> v56
    v53 --> v55
    v53 --> v54
```

**Comment by mribeirodantas** [comment]
Very nice to see the way this is heading to 🤩

File: modules/nextflow/src/main/groovy/nextflow/dag/DAG.groovy [hunk 1]
```diff
@@ -426,7 +426,7 @@ class DAG {
         /**
          * @return The unique name for this node
          */
-        String getName() { "p${getOrder()}" }
+        String getName() { "v${getOrder()}" }
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/dag/DAG.groovy:429]: Why change this prefix?

File: modules/nextflow/src/main/groovy/nextflow/dag/MermaidRenderer.groovy [hunk 2]
```diff
@@ -15,64 +15,490 @@
...
+import nextflow.Global
+import nextflow.Session
 /**
  * Render the DAG using the Mermaid format to the specified file.
  * See https://mermaid-js.github.io/mermaid/#/ for more info.
  *
  * @author Ben Sherman <bentshermann@gmail.com>
  */
+@Slf4j
 class MermaidRenderer implements DagRenderer {
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/dag/MermaidRenderer.groovy:32]: Could be made @CompileStatic?

**Comment by pditommaso** [comment]
Wow, this is cool. Is going to be the default rendering ?

File: modules/nextflow/src/main/groovy/nextflow/dag/DAG.groovy (hunk 1 above)
Comment (bentsherman) [reply modules/nextflow/src/main/groovy/nextflow/dag/DAG.groovy:429]: "v" is for vertex. not sure why it was "p". I guess for process? but then not all vertices are processes

File: modules/nextflow/src/main/groovy/nextflow/dag/MermaidRenderer.groovy (hunk 2 above)
Comment (bentsherman) [reply modules/nextflow/src/main/groovy/nextflow/dag/MermaidRenderer.groovy:35]: done

**Comment by bentsherman** [comment]
Okay, I think we should merge it as is. There are still many things that can be done but would be better as separate in efforts. In particular, improving the input and output labels is likely related to having better input/output definitions for the workflow, so probably best to wait and see.

Summary of future work:

- Extend support to other diagram formats
- Add hyperlinks to processes, operators, etc
- Add interactive expand/collapse to workflow blocks
- Show better labels for inputs and outputs

Also:

> Is it going to be the default rendering ?

The default settings are `verbose = false` and `depth = -1`, so it will hide channel names and operators and it will render the full depth of workflows.

File: modules/nextflow/src/main/groovy/nextflow/dag/DAG.groovy (hunk 1 above)
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/dag/DAG.groovy:429]: Fair enough

**Comment by pditommaso** [comment]
Excellent 👍

**Pull request merged**